### PR TITLE
testament: generic N-fold batching: windows CI 37mn=>16m

### PR DESCRIPTION
**Pull request by timotheecour**
unlike `NIM_TEST_PACKAGES` (which was only batching important_packages individual packages and required hardcoding which package was assigned to which batch, as well as the number of batches), the batching introduced here is generic (just specify number of batches) and adapts the workload based on 2 parameters: number of batches and batch index. batching is fine grained so that each batch empirically takes almost the same time

since windows takes ~3X more time than other azure tests, I'm just batching for windows here

## future work
- [x] remove `NIM_TEST_PACKAGES` and reuse `NIM_TESTAMENT_BATCH` approach for important_packages tests as well (=> https://github.com/nim-lang/Nim/pull/17301)
- [ ] investigate why windows is so slow compared to other OS (in CI at least)
- [ ] be a bit more careful so that other tests are run no more than once (ie, in a single batch); right now the batching only applies to `testament/testament all` which is by far the bottleneck; but some remaining tests could be batched too
- [x] openbsd.yml would be a candidate for 2X batching (=> https://github.com/nim-lang/Nim/pull/14851)

## note
* a good heuristic to batch is to batch randomly, which should give good load balancing; here I'm batching by `hash(filename) mod numBatch` but this could be refined eg using `hash((filename, options))` to avoid having all `tests/gc/gctest` tests (ie with different sub options such as c vs c++) assigned to same batch
* this trades off total cost for faster turnaround, since some of the work (eg building nim) needs to be duplicated in each batch; however, given the fact azure provides N(=10) parallel builds, this is a good tradeoff. 
* the overhead empirically (which could be reduced in future work to be more careful about avoiding some duplicate work) is: (15+15+16)/37 = 1.2

### CODE REVIEW

File: testament/testament.nim [hunk 1]
```diff
@@ -273,12 +273,14 @@ proc addResult(r: var TResults, test: TTest, target: TTarget,
                             expected = expected,
                             given = given)
   r.data.addf("$#\t$#\t$#\t$#", name, expected, given, $success)
+  template disp(msg) =
```
Comment (Araq) [inline testament/testament.nim:276]: Nitpick (don't have to change it), I use ``disp`` for "dispatch" and there is no conditional here. :-)

**Comment by D-Nice** [comment]
I believe this PR broke `./koch tests` for me on Linux/MacOS

```sh
root@1a89339336d9:/tmp/Nim# ./koch tests    
bin/nim cc --opt:speed testament/testament
Hint: used config file '/tmp/Nim/config/nim.cfg' [Conf]
Hint: used config file '/tmp/Nim/config/config.nims' [Conf]
Hint: used config file '/tmp/Nim/testament/testament.nim.cfg' [Conf]
.....................................................................
/tmp/Nim/lib/pure/pegs.nim(1982, 11) Warning: observable stores to 'result.rule' [ObservableStores]
/tmp/Nim/testament/testament.nim(73, 3) Warning: observable stores to 'pegLineError' [ObservableStores]
/tmp/Nim/testament/testament.nim(75, 3) Warning: observable stores to 'pegLineTemplate' [ObservableStores]
/tmp/Nim/testament/testament.nim(80, 3) Warning: observable stores to 'pegOtherError' [ObservableStores]
.
/tmp/Nim/testament/testament.nim(725, 7) Warning: use exitprocs.addExitProc; addQuitProc is deprecated [Deprecated]
/tmp/Nim/testament/categories.nim(52, 6) Hint: 'icTests' is declared but not used [XDeclaredButNotUsed]
CC: stdlib_dollars.nim
CC: testament.nim
Hint:  [Link]
Hint: 123476 lines; 6.810s; 181.77MiB peakmem; Debug build; proj: /tmp/Nim/testament/testament; out: /tmp/Nim/testament/testament [SuccessX]
/tmp/Nim/testament/testament all
progress[all]: 0/121 starting: cat: manyloc
/tmp/Nim/testament/testament --nim:/tmp/Nim/bin/nim --batch: pcat manyloc 
progress[all]: 1/121 starting: cat: system
/tmp/Nim/testament/testament --nim:/tmp/Nim/bin/nim --batch: pcat system 
progress[all]: 2/121 starting: cat: procvar
/tmp/Nim/testament/testament --nim:/tmp/Nim/bin/nim --batch: pcat procvar 
progress[all]: 3/121 starting: cat: distinct
/tmp/Nim/testament/testament --nim:/tmp/Nim/bin/nim --batch: pcat distinct 
/tmp/Nim/testament/testament.nim(774) testament
/tmp/Nim/testament/testament.nim(654) main
/tmp/Nim/lib/system/assertions.nim(30) failedAssertImpl
/tmp/Nim/lib/system/assertions.nim(23) raiseAssert
/tmp/Nim/lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: /tmp/Nim/testament/testament.nim(654, 18) `s.len == 2` ("pcat", @["pcat"]) [AssertionDefect]
progress[all]: 4/121 starting: cat: alias
/tmp/Nim/testament/testament --nim:/tmp/Nim/bin/nim --batch: pcat alias 
/tmp/Nim/testament/testament.nim(774) testament
/tmp/Nim/testament/testament.nim(654) main
/tmp/Nim/lib/system/assertions.nim(30) failedAssertImpl
/tmp/Nim/lib/system/assertions.nim(23) raiseAssert
/tmp/Nim/lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: /tmp/Nim/testament/testament.nim(654, 18) `s.len == 2` ("pcat", @["pcat"]) [AssertionDefect]
progress[all]: 5/121 starting: cat: varres
/tmp/Nim/testament/testament.nim(774) testament
/tmp/Nim/testament/testament.nim(654) main
/tmp/Nim/lib/system/assertions.nim(30) failedAssertImpl
/tmp/Nim/lib/system/assertions.nim(23) raiseAssert
/tmp/Nim/lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: /tmp/Nim/testament/testament.nim(654, 18) `s.len == 2` ("pcat", @["pcat"]) [AssertionDefect]
```

and so on...

fdb3740 works

**Comment by timotheecour** [comment]
@D-Nice please try https://github.com/nim-lang/Nim/pull/14845

**Comment by D-Nice** [comment]
@timotheecour works again following https://github.com/nim-lang/Nim/pull/14845 :+1:

**Pull request merged**